### ISSUE TITLE
[Remote-SSH Bug]: starting "SSH remote connect to host" session hangs indefinitely

### ISSUE THREAD

**Issue by Kilill** [microsoft/vscode-remote-release#7711]
### Is there an existing issue for this bug?

- [X] I have searched the existing issues

### Debugging Steps

- [X] I have followed these debugging steps
- [X] I have tried both values of the `remote.SSH.useLocalServer` setting
- [X] My SSH command works from a local terminal
- [ ] I can connect using the Remote-SSH extension on VS Code when all other VS Code extensions are disabled

### ->

not possible to  connect using remote-ssh extension

### Versions

- VS Code Version:
 ```1.75.0-insider (user setup)
Commit: 11238faea62d570d77afe6edfe05c8b732c44a2b
Date: 2022-12-13T05:22:12.417Z
Electron: 19.1.8
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.19045
Sandboxed: Yes
```
- Remote OS Version:
```
Distributor ID: Ubuntu
Description:    Ubuntu 21.10
Release:        21.10
Codename:       impish
```
- Remote-SSH Extension: 
``` v0.94.0
   and
 v0.95.2022121315
```

### Expected Behavior

the extension should connect to the remote linux system and download the vscode server and connect to it.

### Steps To Reproduce

- local system: 
   Windows 10
   VSCode-insiders
   Remote-ssh extensions installed
   all other extensions disabled

- remote system 
   ubuntu linux system, but in this case it does not matter

### Remote-SSH Log

<details>
<summary>Remote-SSH Log</summary>
<p>

```
[18:09:56.308] Log Level: 1
[18:09:56.377] remote-ssh@0.94.0
[18:09:56.377] win32 x64
[18:09:56.420] Picking SSH host
[18:09:58.197] Selected homeass

```

</p>
</details>

### Anything else?

Trying to open a remote ssh sesion to another linux system correctly opens a new window 
but it hangs indefinitly,

Hitting ``<enter>`` in said window will abort the operations with an "unknown element received" message ( see: "Output on VSCode window for remote" below).

It is possible to ssh into the remote system from a local cmd  line / terminal / powershell or  a VSCode terminal.

Opening a local "cmd" and running the command line:
``` 
C:\WINDOWS\System32\cmd.exe /c ( type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-534327826.sh" | ssh -T -D 60759 homeass bash )
``` 
fails with  bash complaining about the trailing ")"
removing both parenthesis and reruning as:
``` 
C:\WINDOWS\System32\cmd.exe /c  type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-534327826.sh" | ssh -T -D 60759 homeass bash 
``` 
works, ie , the ssh connection is established the bash script is executed on the remote, the vscode server is downloaded and started.

<details>
<summary>Remote - SSH Log</summary>
<p>

```
<details>
<summary>Pty Host Log</summary>
<p>

```
2022-12-14 18:10:00.192 [warning] Shell integration cannot be enabled for executable "C:\WINDOWS\System32\cmd.exe" and args /c (type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-123494529.sh" | ssh -T -D 58038 homeass bash) & exit /b 0
```

</p>
</details>
```

</p>
</details>
<details>
<summary>Pty Host Log</summary>
<p>

```
2022-12-14 18:10:00.192 [warning] Shell integration cannot be enabled for executable "C:\WINDOWS\System32\cmd.exe" and args /c (type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-123494529.sh" | ssh -T -D 58038 homeass bash) & exit /b 0
```

</p>
</details>
<details>
<summary>"Output on VSCode window for remote</summary>
<p>

```
[18:24:54.627] Log Level: 1
[18:24:54.632] remote-ssh@0.95.2022121315
[18:24:54.632] win32 x64
[18:24:54.641] SSH Resolver called for "ssh-remote+homeass", attempt 1
[18:24:54.642] "remote.SSH.useLocalServer": false
[18:24:54.643] "remote.SSH.showLoginTerminal": true
[18:24:54.644] "remote.SSH.remotePlatform": {"homeass.killhome.itr.no":"linux","homeass":"linux"}
[18:24:54.644] "remote.SSH.path": undefined
[18:24:54.644] "remote.SSH.configFile": undefined
[18:24:54.645] "remote.SSH.useFlock": true
[18:24:54.645] "remote.SSH.lockfilesInTmp": true
[18:24:54.646] "remote.SSH.localServerDownload": auto
[18:24:54.646] "remote.SSH.remoteServerListenOnSocket": false
[18:24:54.647] "remote.SSH.showLoginTerminal": true
[18:24:54.647] "remote.SSH.defaultExtensions": []
[18:24:54.648] "remote.SSH.loglevel": 1
[18:24:54.649] "remote.SSH.enableDynamicForwarding": true
[18:24:54.649] "remote.SSH.enableRemoteCommand": false
[18:24:54.650] "remote.SSH.serverPickPortsFromRange": {}
[18:24:54.651] "remote.SSH.serverInstallPath": {}
[18:24:54.664] SSH Resolver called for host: homeass
[18:24:54.664] Setting up SSH remote "homeass"
[18:24:54.669] Using commit id "11238faea62d570d77afe6edfe05c8b732c44a2b" and quality "insider" for server
[18:24:54.674] Install and start server if needed
[18:24:54.677] Checking ssh with "ssh -V"
[18:24:54.728] > OpenSSH_for_Windows_8.1p1, LibreSSL 3.0.2

[18:24:54.735] Running script with connection command: ssh -T -D 59232 homeass bash
[18:24:54.738] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[18:24:55.231] "install" terminal received data: "[2J[m[H]0;C:\WINDOWS\System32\cmd.exe[?25h"
[18:24:55.231] Got some output, clearing connection timeout
[18:24:55.241] "install" terminal received data: "[25l[K
[K
[K
[K
[K
[K
[K
[K
[K
[K
[K[H[?25h"
[18:26:18.494] "install" terminal received data: ""
[18:26:18.510] "install" terminal received data: "An unknown element "" was received. This can happen if the remote process closed or ended abnormally."
[18:26:19.783] "install" terminal command done
[18:26:19.783] Install terminal quit with output: An unknown element "" was received. This can happen if the remote process closed or ended abnormally.
[18:26:19.784] Received install output: An unknown element "" was received. This can happen if the remote process closed or ended abnormally.
[18:26:19.784] Stopped parsing output early. Remaining text: An unknown element "" was received. This can happen if the remote process closed or ended abnormally.
[18:26:19.784] Failed to parse remote port from server output
[18:26:19.786] Resolver error: Error: 
	at g.Create (c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:583926)
	at t.handleInstallOutput (c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:582540)
	at t.tryInstall (c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:680478)
	at async c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:642848
	at async t.withShowDetailsEvent (c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:646166)
	at async t.resolve (c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:643898)
	at async c:\Users\kill\.vscode-insiders\extensions\ms-vscode-remote.remote-ssh-0.95.2022121315\out\extension.js:1:723042
[18:26:19.789] TELEMETRY: {"eventName":"resolver","properties":{"osReleaseId":"","arch":"","askedPw":"0","askedPassphrase":"0","asked2fa":"0","askedHostKey":"0","remoteInConfigFile":"1","gotUnrecognizedPrompt":"0","dynamicForwarding":"1","localServer":"0","didLocalDownload":"0","installUnpackCode":"0","outcome":"failure","reason":"UnparsableOutput","exitCodeLabel":""},"measures":{"resolveAttempts":1,"retries":1,"timing.totalResolveTime":85147,"timing.preSshTime":100,"timing.scriptTime":85046}}
[18:26:19.793] ------
```

</p>
</details>

**Comment by eleanorjboyd** [comment]
Hello @Kilill, thank you for the in-depth report! Has the extension worked for you in the past? It looks to me like there is a firewall issue on your server that is truncating the install script. Do you have any firewall or company settings on your remote machine that could be causing this?

**Comment by Kilill** [comment]
Not realy tested this feature before.
Nope, Both systems are on the same network, none of the have any firewalls running.
Besides, that would have also stopped it from also executing from a cmd prompt, but that as i mentioned works just fine.
Its only trying to execute the "Connect to remote.." from vscode that fails.

**Comment by eleanorjboyd** [comment]
Hello! On further investigation it looks like the error you are getting ` "install" terminal received data: "An unknown element "" was received.`, is one that comes from SSH and not from our extension. What outcome did you get when you tried number 3 on the debugging doc `Now try running this command but add echo "echo hello" | to the beginning of your SSH command to confirm that we can execute a script on your remote by piping it into SSH`?

**Comment by Kilill** [comment]
As i mentioned originally running the command from a normal "command prompt" works just fine

here is a complete run of the same:
<details>
<summary>C:\Users\kill>C:\WINDOWS\System32\cmd.exe /c  type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-534327826.sh" | ssh -T -D 60759 homeass bash</summary>
<p>

```
Microsoft Windows [Version 10.0.19045.2311]
(c) Microsoft Corporation. All rights reserved.

C:\Users\kill>C:\WINDOWS\System32\cmd.exe /c  type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-534327826.sh" | ssh -T -D 60759 homeass bash
c81234cfffcd: running
Acquiring lock on /run/user/1000/vscode-remote-lock.kill.11238faea62d570d77afe6edfe05c8b732c44a2b
Installing to /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b...
c81234cfffcd%%1%%
Downloading with wget
Download complete
c81234cfffcd%%2%%
tar --version:
tar (GNU tar) 1.34
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John Gilmore and Jay Fenlason.
Checking /home/kill/.vscode-server-insiders/.11238faea62d570d77afe6edfe05c8b732c44a2b.log and /home/kill/.vscode-server-insiders/.11238faea62d570d77afe6edfe05c8b732c44a2b.pid for a running server
Running ssh connection command... /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b/bin/code-server-insiders --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all   &> "/home/kill/.vscode-server-insiders/.11238faea62d570d77afe6edfe05c8b732c44a2b.log" < /dev/null
printenv:
    SHELL=/bin/bash
    SSH_AUTH_SOCK=/tmp/ssh-gez8bQrVXY/agent.30528
    PWD=/home/kill
    LOGNAME=kill
    XDG_SESSION_TYPE=tty
    MOTD_SHOWN=pam
    HOME=/home/kill
    LANG=en_US.UTF-8
    VSCODE_AGENT_FOLDER=/home/kill/.vscode-server-insiders
    SSH_CONNECTION=192.168.127.3 57314 192.168.127.54 22
    XDG_SESSION_CLASS=user
    USER=kill
    SHLVL=0
    XDG_SESSION_ID=14
    XDG_RUNTIME_DIR=/run/user/1000
    SSH_CLIENT=192.168.127.3 57314 22
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
    DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
    OLDPWD=/home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b
    _=/usr/bin/printenv
rm: cannot remove '/home/kill/.vscode-server-insiders/.11238faea62d570d77afe6edfe05c8b732c44a2b.token': No such file or directory
Spawned remote server: 30607
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...
Waiting for server log...

*
* Visual Studio Code Server
*
* By using the software, you agree to
* the Visual Studio Code Server License Terms (https://aka.ms/vscode-server-license) and
* the Microsoft Privacy Statement (https://privacy.microsoft.com/en-US/privacystatement).
*

c81234cfffcd: start
SSH_AUTH_SOCK==/tmp/ssh-gez8bQrVXY/agent.30528==
DISPLAY====
webUiAccessToken====
listeningOn==36595==
osReleaseId==ubuntu==
arch==x86_64==
tmpDir==/run/user/1000==
platform==linux==
unpackResult==success==
didLocalDownload==0==
downloadTime==7730==
installTime==9306==
extInstallTime====
serverStartTime==1693==
connectionToken==096777df-6ea8-4fc4-84fc-a93b35e763b6==
c81234cfffcd: end
```

</p>
</details>

Checking the remote system the .``` .vscode-server-insiders``` directory is created, and the server has been started:
``` 
$ ps ax | grep vscode
30617 ?        Sl     0:03 /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b/node /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b/out/server-main.js --start-server --host=127.0.0.1 --accept-server-license-terms --enable-remote-auto-shutdown --port=0 --telemetry-level all --connection-token-file /home/kill/.vscode-server-insiders/.11238faea62d570d77afe6edfe05c8b732c44a2b.token
30828 ?        Sl     0:01 /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b/node /home/kill/.vscode-server-insiders/bin/11238faea62d570d77afe6edfe05c8b732c44a2b/out/bootstrap-fork --type=ptyHost --logsPath /home/kill/.vscode-server-insiders/data/logs/20221215T203819
```

**Comment by CanyonCasa** [comment]
I am experiencing a similar behavior as of the last few days... I installed Remote-SSH v0.94.0 about 5 weeks ago on a Windows 10 machine and have been using it to access an AWS Lightsail instance running Debian 11 without incident. But over the last few days when trying to reconnect from a session dropout, it occasionally hangs indefinitely. It locks up the (remote) AWS instance, including any other (Putty) SSH sessions, and completely drains burst capacity (meaning it is churning away on something at max CPU rate) requiring me to stop and restart the instance to regain control and recover. VSCode (1.74.1) also locks up that I can only recover by killing it and restarting.

**Comment by eleanorjboyd** [comment]
Hello @CanyonCasa! Could you submit a new issue with your bug? I am not sure if your problem is exactly related and seeing your logs would be helpful thanks!

**Comment by eleanorjboyd** [comment]
@Kilill, Hello! Thank you for those logs those are helpful! Do you have anything in your bash.rc or any remote command? The connection seems to disconnect before we can start running our install script and those are the two actions we take in between authentication and our install script.

**Comment by roblourens** [comment]
I'm stumped on this one, I have no clue what the difference is between when you ran the command manually and when we run the same command. The only reference to that error online that I see is in powershell, but we are only running the command in cmd.exe.

You said you ran it with `useLocalServer` enabled, did you get the same error message there or a different one?

**Comment by Kilill** [comment]
Ah!! thanks for the hint: "_The only reference to that error online that I see is in **powershell**, but we are only running the command in **cmd.exe**._"
I have %COMSPEC% set to _powershell.exe_ instead of the default _cmd.exe_. Forgot all about that
changing it back to _cmd.exe_ solved the problem.

(background: platformio that's python 3.8 based, also calls up a system shell via %COMPSEC%. If the path contains a networkdrive letter, python stubbornly expands the path to a full UNC path, which makes cmd.exe barf)

**Comment by roblourens** [comment]
That's still a bit odd because we specifically use `cmd.exe`. I don't know what ends up checking comspec. And I don't know why powershell failed with that message. Glad that fixed it for you though.

**Comment by eleanorjboyd** [comment]
@roblourens, does this seem like something we should look into in order to surface a better error message. Not sure how common/in scope this issue is. Thanks

**Comment by roblourens** [comment]
I don't have a good understanding of what's going on here, so I can't really say.

**Comment by Kilill** [comment]
for the heck of i  did some investigating.
Retried the command under power shell and it fails:
```
PS C:\Users\kill> C:\WINDOWS\System32\cmd.exe /c  type "C:\Users\kill\AppData\Local\Temp\vscode-linux-multi-line-command-homeass-977240591.sh" | ssh -T -D 60759 homeass bash
bash: line 1: $'\r': command not found
0dcbef0573fa: running
bash: line 3: $'\r': command not found
bash: line 16: $'\r': command not found
bash: line 18: $'\r': command not found
bash: line 27: $'\r': command not found
bash: line 33: $'\r': command not found
bash: line 34: syntax error near unexpected token `$'{\r''
'ash: line 34: `fail_with_exitcode() {
```
the same happens if dropping the initial `C:\WINDOWS\System32\cmd.exe /c` ie letting power shell do the `type /c` command

It seems that powershell just ignores the call to cmd.exe and run's the `type /c` itself, but `type` in poweshell is and alias for the `Get-Content` cmdlet that does something else than   the old DOS `type` , it reads the file as an arrays of strings and prints them out with an terminating `\r` , which ofc. confuses bash to no end....
This can be verified by replacing the `type /c` with `Get-Content`, it fails with the same output.

The correct way would be to use `Get-Content -Raw` that does not alter the stream,
but then one would have  to figure out if it is actually `powershell` or `cmd` thats running.

**Sugestion:** use internal vscode code that spawns the ssh comand , reads the file and pipes into `ssh` or even better , don't create the file in the first place and just write the commands directly to the ssh pipe...

**Comment by roblourens** [comment]
> use internal vscode code that spawns the ssh comand , reads the file and pipes into ssh or even better , don't create the file in the first place and just write the commands directly to the ssh pipe...

Yeah, that's what useLocalServer=true does, but has other issues on windows so is disabled by default.

> It seems that powershell just ignores the call to cmd.exe and run's the type /c itself

Any idea why that is happening? That is definitely wrong, and seems to just be a problem on your system. Can you tell whether that happens for other `cmd.exe /c ...` commands?
